[iOS][team-09] MVVM구조 리팩토링 및 테스트 코드 작성

JasonLee0223 opened the pull request:
안녕하세요! 만사🙋🏻 
시간이 어느덧 벌써 금요일이 되었네요..
Sol: 이번주에도 MVVM 구조에 대한 학습을 주로 진행했고, 테스트 코드를 작성해보는 정도의 진행을 했습니다.
이전 리뷰에 대한 반영을 진행하면서 대부분의 진행을 하여, PR에 포함된 커밋의 양이 많지는 않습니다ㅠㅠ
리뷰 잘 부탁드리겠습니다!

Jason: Sol과 같이 저도 MVVM 구조에 대해 학습을 많이 해보았고 처음 MVC -> MVVM으로 구조를 변경하다보니 많이 어색하고
익숙치않아 적용하는데 저는 조금 더 오래걸렸네요 ㅠㅠ 하여 지난번 수요일에는 코드가 많지 않아서 패스하고 진행하였습니다 ㅠㅠ
리뷰 잘 부탁드립니다!! 🙇🏻 


### **🧑🏼‍💻 작업 목록**
- [x] IssueView MVVM 구조 개선
- [x] IssueViewModel 구현
- [x] API 연결하여 데이터 출력 확인
- [x] SignInScene의 MVVM 구조 개선
- [x] EditingIssueScene의 구조를 MVVM 구조로 변경
- [x] EditingIssueScene에서 `저장` 기능을 위해 네트워크와 연결
- [x] EditingIssueScene에 대한 테스트 코드 작성 및 SignInScene의 테스트 코드 수정
- [x] IssueCell이 여러 레이블을 보여주도록 뷰 수정


### **💭 학습 키워드**
`MVVM`
`URLRequest`
`URLComponent`
`URLQueryItem`
`Binding`

### **🤔 고민과 해결**
- 기존에 SignInViewModel에서, actions를 사용하는 부분이 어색하다고 느껴졌습니다.
    
    기존에는 실행하려는 Action의 속성 값(클로저)를 VC에서 설정해주고, View의 로그인버튼이 눌리면 ViewModel의 Input에 해당하는 메소드에서 앞서 설정해준 Action을 실행시켜주는 흐름이었습니다.
    
    이렇게 진행될 경우, 그냥 VC에서 해당 버튼에 대한 액션 로직을 실행시켜주는 것(@IBAction처럼)과 다름이 없다고 생각해 이 부분이 어색하다고 느껴졌습니다.
    
    그래서 Action에 관한 내용을 제거하고 아래의 흐름으로 ViewModel의 동작을 수정해보았습니다.
    > 1. 어떤 버튼이 눌렸을 때 그저 ViewModel의 Input에 해당하는 메소드를 호출한다.
    > 2. 해당 Input 메소드에서는 Observable 속성의 value를 변경시켜준다.
    > 3. value가 변경된 Observable 속성은 VC에서 지정해준 block을 **자동으로** 실행한다.

    이렇게 변경하는 것이 좀 더 ViewModel의 역할과 잘 맞는다 생각하여, 아래와 같이 코드를 수정했습니다.
    ```swift
    protocol EditingIssueViewModelInput {
        // 위에서 말한 ViewModel의 Input에 해당하는 메소드
        func didTouchCancel() 
    }

    protocol EditingIssueViewModelOutput {
        // 위에서 말한 Observable 속성
        var cancelButtonState: Observable<Bool> { get } 
    }

    protocol EditingIssueViewModelProtocol: EditingIssueViewModelInput, EditingIssueViewModelOutput { }

    final class EditingIssueViewModel: EditingIssueViewModelProtocol {
        private(set) var cancelButtonState: Observable<Bool> = Observable(false)

        // 2. Input 메소드 내에서는 Observable 속성의 value만 변경시켜준다.
        func didTouchCancel() { 
            cancelButtonState.value = true
        }
    }


    final class EditingIssueViewController {
        func bind(to viewModel: EditingIssueViewModelProtocol) {
            // VC에서 Observable 속성의 value가 변경될 때 실행될 block을 지정 (3에서 사용됨)
            viewModel.cancelButtonState.bind(on: self) { [weak self] _ in 
                self?.popViewController()
            }
        
		// 1. 버튼과 ViewModelInput 메소드 연결
            navigationItems.cancelButton.addAction(UIAction { [weak self] _ in
                self?.viewModel.didTouchCancel()
            }, for: .touchUpInside)
        }
    }
    ```

### **💬  실행 화면**
|변경된 IssueView|EditingIssue 저장버튼|API로 받아온 Issue화면|
|---|---|---|
|![SS 2022-06-24 PM 02 12 49](https://user-images.githubusercontent.com/92504186/175466800-5dc3f6da-125f-445e-9c8d-f8d311b3ff16.jpg)|![SS 2022-06-24 PM 02 31 06](https://user-images.githubusercontent.com/92504186/175468810-c8f732b6-b685-405e-811b-046e8536739b.gif)|![](https://i.imgur.com/X8MhxT9.png)
|

### **🙋🏻 질문 거리**
- VC의 책임 분리를 위해 CollectionViewDataSource를 VC로부터 분리했었습니다.
 
    그런데 DataSource에서 사용할 `items: [IssueItem]` 프로퍼티와 ViewModel이 Observe해야하는 데이터(ex. `items: Observable<[IssueItem]>`)가 같기 때문에, DataSource에서 ViewModel의 해당 프로퍼티를 참조해서 셀의 configuration 결정 및 셀의 개수 등을 결정해주는게 좋을 것 같다고 생각했습니다.
     
    그래서 DataSource에서도 VC와 같은 ViewModel을 참조하려고 하는데, DataSource와 같은 객체에서 ViewModel을 갖는게 어색할까요?

JasonLee0223 committed:
refactor: issueItem CodingKey 작성

JasonLee0223 committed:
feat: IssueViewModel 생성

JasonLee0223 committed:
refactor: IssueDatsource issueViewmodel count로 변경

JasonLee0223 committed:
refactor: IssueItem condingKey 맞춤

Hansolkkim committed:
feat: EditingIssueManager 구현 (bugpigg/issue-tracker#33)

- EditingIssueScene의 usecase로 사용할 EditingIssueManager 구현
- EditingIssueManager에서 네트워크 통신시 사용할 EditingIssueNetworkTarget 구현
- POST에 대응하기 위해 NetworkService 메소드 수정

Hansolkkim committed:
feat: save버튼 및 에러에 대한 Observable 프로퍼티 구현 및 usecase 프로퍼티 추가 (bugpigg/issue-tracker#33)

Hansolkkim committed:
test: EditingIssueManager 객체에 대한 테스트 코드 작성 (bugpigg/issue-tracker#33)

- 테스트를 위해 해당 객체의 urlSession 속성을 설정해줄 필요가 있어, 해당 속성을 변수로 바꾸고, 해당 속성을 설정해주는 메소드 추가 구현

Hansolkkim committed:
refactor: SignInViewModel 구조 개선 (bugpigg/issue-tracker#32)

- Actions를 따로 정의하고 실행되게 하는 부분은 MVVM 구조와 맞지 않다고 판단하여 SignInViewModelActions 제거
- SignInViewModelOutput에 "gitHubSignInURL: Observable<URL?>" 속성 추가

Hansolkkim committed:
test: SignInViewModel에 대한 테스트 코드 수정 (bugpigg/issue-tracker#32)

- SignInViewModel 변경에 따른 테스트 코드 수정
- SignInManager의 urlSession 속성을 변경시켜줄 수 있도록 코드 수정
- 테스트에 있는 불필요한 부분 제거 및 에러 수정

Hansolkkim committed:
Merge remote-tracking branch 'origin/iOS-refactor/SignInScene' into iOS

Hansolkkim committed:
feat: IssueCell에 여러 레이블을 추가할 수 있도록 뷰 수정

JasonLee0223 committed:
Refactor: IssueManager 생성 및 구현

JasonLee0223 committed:
feat: IssueNetworkTarget 생성 및 구현

JasonLee0223 committed:
refactor: IssueViewController에 ViewModel과 View를 Bind하는 메서드 구현

JasonLee0223 committed:
refactor: API에서 변수명 변경사항 적용

JasonLee0223 committed:
refactor: ViewModel의 Observable 적용 및 Configure 수정

JasonLee0223 committed:
Merge branch 'iOS-#11' into iOS

# Conflicts:
#	iOS/IssueTracker/IssueTracker.xcodeproj/project.pbxproj
#	iOS/IssueTracker/IssueTracker/Issue/View/IssueCollectionViewCell.swift

iOS-#11 충돌 해결 후 Merge

JasonLee0223 committed:
refactor: IssueCollectionViewCell의 Configure complict해결 후 남아있는 오류 해결

Hansolkkim committed:
chore: 폴더 정리

ITzombietux commented on iOS/IssueTracker/IssueTracker/Common/Model/NetworkService.swift:
```diff
@@ -46,6 +46,11 @@ private extension NetworkService {
         if let url = components.url {
             var request = URLRequest(url: url)
             request.httpMethod = target.method
+            
+            if let target = target as? EditingIssueNetworkTarget {
+                request.httpBody = target.body
+                request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+            }
```
if 안에 또 if 가 생기니 가독성이 약간 떨어지는게 있을거 같아요!!!

Hansolkkim replied:
안의 if문을 없애기 위해, NetworkTargetProtocol에 Data?타입의 body라는 프로퍼티를 정의해주도록 하여, body가 없는 case들은 nil을 반환하도록 변경해보겠습니다 ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/IssueCollectionViewDataSource.swift:
```diff
@@ -9,34 +9,26 @@ import UIKit
+        self.issueViewModel = issueViewModel
+    }
     
     func numberOfSections(in collectionView: UICollectionView) -> Int {
         return 1
     }
```
numberOfSections 메서드를 선언하지 않으면 디폴트로 1을 리턴하는데 메서드를 지워도 되지 않을까요?ㅎㅎㅎ

JasonLee0223 replied:
넵! Default로 적용이 되는 점 반영하였습니다!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/LabelFactory.swift:
```diff
@@ -0,0 +1,23 @@
+import UIKit
+
+struct LabelFactory {
+    static func createLael(_ label: Label) -> PaddingLabel {
```
오타가 있는거 같아요!! 👀

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/EditingIssueManager.swift:
```diff
@@ -0,0 +1,36 @@
+    mutating func setURLSession(_ session: URLSessionProtocol)
+}
+
+struct EditingIssueManager {
```
EditingIssueManager를 struct로 만드신 이유가 있으실까요?!

Hansolkkim replied:
굳이 class를 사용하지 않아도 괜찮은 객체라 생각해서 struct로 만들었습니다!

값의 변경이 계속 해서 발생한다기보다는, viewModel에서 필요할 때만 호출해서 원하는 값을 받아오도록 해주는 객체라고 생각해서 class일 필요가 없다고 생각했습니다!

혹시 usecase에 해당하는 객체는 class가 되는게 더 일반적일까요?!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/EditingIssueManager.swift:
```diff
@@ -0,0 +1,36 @@
+            completion: completion)
+    }
+
+    mutating func setURLSession(_ session: URLSessionProtocol) {
+        urlSession = session
+    }
```
setter를 사용하지 않는 방법이 있을까요?!?! ㅎㅎ 이게 메서드이긴 하지만 urlSession 속성을 var 로 만든 것과 차이가 없다라고 생각이 되기도 하는데요!!

Hansolkkim replied:
테스트를 할 때, 테스트 객체 초기화 시 EditingIssueManagable 타입의 구체 타입을 만들고, 테스트 메소드에서 해당 타입의 urlSession을 각 메소드에서 바꾸어주어야 한다고 생각해서 setter를 만들었습니다..

 그런데, 생각해보니 어짜피 mockAPI에 대한 테스트를 진행할거라 init 시에 urlSession을 지정해주는 방법을 사용할 수 있을 것 같습니다!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/EditingIssueNetworkTarget.swift:
```diff
@@ -0,0 +1,61 @@
+    case sendIssue(newIssue: IssueItem)
+}
+
+extension EditingIssueNetworkTarget: NetworkTargetProtocol {
+    var url: String {
+        return baseURL + path
+    }
+    
+    var queryItem: [URLQueryItem]? {
+        switch self {
+        case .sendIssue:
+            return nil
+        }
+    }
+    
+    var method: String? {
+        switch self {
+        case .sendIssue:
+            return "POST"
+        }
+    }
+
+    var body: Data? {
+        switch self {
+        case .sendIssue(let newIssue):
+            let parameter: [String: Any] = [
+                "title": newIssue.title,
+                "content": newIssue.content,
+                "milestoneTitle": newIssue.milestoneName,
+                "label": ["0"],
+                "assignee": ["0"]
+            ]
+
+            let body = try? JSONSerialization.data(withJSONObject: parameter)
+            return body
+        }
+    }
+}
```
EditingIssueNetworkTarget 말고 다른 네트워크 타겟이 생기거나 또 다른 호스트가 생긴다고 생각해보면 이 코드가 비슷한 형태로 중복이 될거 같은데 재사용할 수 있는 방법이 있지 않을까?? 하는 생각에 잠겨봅니다..!

Hansolkkim replied:
음.. 어떤 말씀이신지 제대로 이해를 못 했습니다ㅠㅠ
혹시 조금만 힌트를 주실 수 있을까요..?🤔

ITzombietux replied:
앗! 이 부분은 그냥 넘어가도록 하시죠!! ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/IssueItem.swift:
```diff
@@ -7,15 +7,22 @@
+// MARK: - Label
+struct Label: Codable {
+    let title, backgroundColor: String
+    let darkMode: Bool
```
darkMode의 데이터 타입이 Bool이라면 isDarkMode가 좋지 않을까요?!?!

JasonLee0223 replied:
당시에 Codable Error가 계속 발생하여 API에 맞춰서 작성하다보니 네이밍이 조금 이상해졌네요😅
아래와 같이 변경해보았습니다!
```Swift
struct Label: Codable {
    let title, backgroundColor: String
    let darkMode: Bool
    
    enum LabelCodingKeys: String, CodingKey {
        case title, backgroundColor
        case darkMode = "isDarkMode"
    }
}
```

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/IssueNetworkTarget.swift:
```diff
@@ -0,0 +1,58 @@
+}
+
+extension IssueNetworkTarget: NetworkTargetProtocol {
+    var queryItem: [URLQueryItem]? {
+        switch self {
+        case .issuesList:
+            return nil
+        case .issueListDetail:
+            return nil
+        case .createIssue:
+            return nil
+        case .deleteIssue:
+            return nil
+        }
+    }
+
+    var url: String {
+        return self.baseURL + self.path
+    }
+
+    var method: String? {
+        switch self {
+        case .issuesList, .issueListDetail:
+            return "GET"
+        case .createIssue:
+            return "POST"
+        case .deleteIssue:
+            return "DEL"
+        }
+    }
+}
+
```
위에서 언급헀던 비슷한 형태의 코드가 발생을 했네요!! ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/View/IssueViewComponents/IssueCollectionView.swift:
```diff
@@ -59,6 +59,10 @@ final class IssueCollectionView: UIView {
         addNewIssueButton.addAction(action, for: .touchUpInside)
     }
     
+    func update() {
```
메서드 네이밍과 내용이 뭔가 매칭이 잘 안되는거 같기도 하네요! 이후에 메서드 안에 또 다른 메서드가 들어갈까요!? 조금 더 명확한 의미전달이 되면 더더 좋을거 같아요!

JasonLee0223 replied:
`updateIssueView` 라는 네이밍으로 변경하여 View의 업데이트 되는 내용만 reload된다는 의미로 변경해보았습니다!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/View/IssueViewComponents/IssueCollectionViewCell.swift:
```diff
@@ -48,22 +48,25 @@ final class IssueCollectionViewCell: UICollectionViewCell {
+        self.issueDescription.text = viewModel.content
+        self.milestoneName.text = viewModel.milestoneName
+
+        viewModel.labels.forEach { label in
```
forEach를 configure메서드 안에서 하는 것이 아니라 viewModel.forEachLabels 같은 메서드를 제공해주는건 어떨까요?!?!

Hansolkkim replied:
혹시 말씀하신 메서드가 어떤 동작을 하게 되는 메서드를 의미하는지 알 수 있을까요?
어떤 이유에서 메서드를 제공하는게 적절하다고 생각하신지도 궁금합니다!

제가 생각해봤을 때, `[PaddingLabel] 을 반환하는 메서드`, 또는 `handler를 받아서 IssueItem 타입 내에서 처리할 클로저를 받는 메서드` 두 경우가 생각되는데, 
첫 번째 방법을 사용하면, 반환받은 [PaddingLabel]에 대한 forEach문이 또 필요할 것 같고,
두 번째 방법을 사용하면 forEach문과 다르지 않은 형태를 가질 것 같습니다.

ITzombietux replied:
forEach같은 경우는 큰 문제가 없어 보이긴 하지만! 이렇게 생각을 해보는건 어떨까요?!?!
labels.count 나 labels.isEmpty 같은 동작 역시 할 수 있어보이는데요! 어레이 타입에 제한을 두면 어떨까? 하는 제 의견이었는데 사실 이 로직에선 불필요해 보이기도 합니다!! ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/View/IssueViewController.swift:
```diff
@@ -10,9 +10,22 @@ import UIKit
+    private var dataSource: IssueCollectionViewDataSource
     private var searchController = UISearchController(searchResultsController: nil)
     
+    let issueViewModel: IssueViewModel
```
IssueViewModelOutput을 인스턴스로 갖고 있는건 어떨까요?!?! ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/ViewModel/EditingIssueViewModel.swift:
```diff
@@ -10,23 +10,34 @@ import UIKit
 }
 
 protocol EditingIssueViewModelOutput {
     var titleText: Observable<String> { get }
     var contentText: Observable<String> { get }
     var cancelButtonState: Observable<Bool> { get }
+    var saveButtonState: Observable<Bool> { get }
     var segmentIndex: Observable<Int> { get }
+    var error: Observable<String> { get }
```
굉~장히 디테일하네요! ㅎㅎㅎ 좋습니다.!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/ViewModel/EditingIssueViewModel.swift:
```diff
@@ -10,23 +10,34 @@ import UIKit
 protocol EditingIssueViewModelProtocol: EditingIssueViewModelInput, EditingIssueViewModelOutput { }
 
 final class EditingIssueViewModel: EditingIssueViewModelProtocol {
+    private let useCase: EditingIssueManagable
```
useCase라는 네이밍보다는 객체를 나타내는 의미가 좀 더 좋지 않을까 생각해봅니다!!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/IssueManager.swift:
```diff
@@ -0,0 +1,52 @@
+
+import Foundation
+
+struct IssueManager {
```
이 객체도 IssueManageable 같은 객체로 의존성 역전을 하면 좋지 않을까 한번 생각해봅니다! struct로 만드셨네요!

ITzombietux commented on iOS/IssueTracker/IssueTracker/SignIn/View/SignInViewController.swift:
```diff
@@ -32,15 +32,15 @@ final class SignInViewController: UIViewController {
     func bind(to viewModel: SignInViewModelProtocol) {
         signInView.setGitHubSignInButtonAction(UIAction { [weak self] _ in
-            self?.viewModel.didSelect()
+            self?.viewModel.didTouchGitHubSignIn()
```
가독성이 좋아졌네요! ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/IssueCollectionViewDataSource.swift:
```diff
@@ -9,34 +9,26 @@ import UIKit
-            .init(title: "iOS", backgroundColor: "#20bb87")
-        ])
-    ]
+    let issueViewModel: IssueViewModel
```
뷰모델은 모델의 변경을 감지하고, 뷰가 사용할 수 있는 데이터로 변환하고, UI관련 코드는 없다는 특징이 있자나요?~!! 그런걸로 생각해봤을 때 데이터소스라는 UI에서 "모델의 변경을 감지하는 행동(Input)", "뷰가 사용할 수 있는 데이터로 변환(output)" 하는 동작을 하지 않고 변경된 UI 적용을 위해서 사용한다면 괜찮지도 않을까? 라고 생각이 들기도 하고, 변경된 UI만 적용시킬 것이라면 굳이 뷰모델 전부를 인스턴스화 할 필요가 있을까? 싶기도 합니다. !!!

Hansolkkim replied:
답변 감사합니다!

그래서 DataSource에서 issueViewModel 말고, Observable<[IssueItem]> 타입의 속성을 가지고 있을지,
아니면 [IssueItem] 타입의 속성을 가지고 있을지 고민해봤습니다!

그런데 [IssueItem] 속성을 가지고 있는 경우에는, IssueItem은 struct로 구현된 타입이라, DataSource 객체가 초기화될 때 viewModel의 [IssueItem] 속성을 복사해서 그대로만 가지고 있게 된다는 문제가 있어, 
[IssueItem]이 네트워크 통신 후 변경되는 경우에는 이를 DataSource 객체에서는 알지 못 하게 되는 경우가 발생합니다.

그래서 DataSource에서 Observable<[IssueItem]> 타입의 속성을 초기화시에 viewModel로부터 받아, 네트워크 통신 후에도 viewModel과 같은 [IssueItem]을 알고 있는게 더 적합하다고 느껴졌습니다.

그래서 `issueViewModel: IssueViewModel` 속성을 `data: Observable<[IssuItem]>` 속성으로 바꾸면, 뷰모델 전부를 인스턴스화할 필요도 없고, 네트워크 통신 후 변경된 [IssueItem]을 DataSource 객체 또한 알 수 있도록 할 수 있을 것 같습니다!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/EditingIssueManager.swift:
```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+protocol EditingIssueManagable {
+    typealias IdData = [String: Int]
```
IdData라는 네이밍이 사실 조금 이해하기 어렵지 않나?! 생각이 듭니다!! 제가 이해하지 못한 것일 수도 있습니다!!

Hansolkkim replied:
네이밍에 너무 많은 시간이 들어, 대충 지었던 이름이었던 것 같습니다..ㅎㅎ

해당 네이밍의 타입은 Issue를 저장한다는 POST 요청을 API로 보내면, 서버에서는 해당 Issue에 대한 id를 반환해주는데 이를 Dictionary로 받아올 때 사용하는 DTO 타입입니다. 그래서 이런 이름을 지었었는데, 조금 더 생각이 필요할 것 같습니다! 생각하고 수정해서 반영해보겠습니다ㅎㅎ

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/Model/EditingIssueNetworkTarget.swift:
```diff
@@ -0,0 +1,61 @@
+
+import Foundation
+
+enum EditingIssueNetworkTarget {
+    case sendIssue(newIssue: IssueItem)
+}
```
enum 네이밍이 Editting으로 시작을 하고 있는데 newIssue 라는 파라미터 이름은 좀 어색해 보이네요! "수정된" 이럼 느낌은 어떨까요?!?!

Hansolkkim replied:
새로운 이슈를 생성하는 뷰와, 이슈를 수정하는 뷰를 동일한 VC를 사용하라는 요구조건이 있어 네이밍에 애매한 부분이 있었던것 같습니다..ㅠㅠ

그런데 아직 수정하는 경우에 대한 로직을 구현하지 못 해 해당 NetworkTarget에도 case를 나누지 못 했습니다,,
네이밍에 조금 더 고민을 해보고 반영해보곘습니다!

ITzombietux commented on iOS/IssueTracker/IssueTracker/Issue/IssueCollectionViewDataSource.swift:
```diff
@@ -9,34 +9,26 @@ import UIKit
             return UICollectionViewCell()
         }
-        let issueEntity = data[indexPath.item]
+        let issueEntity = issueViewModel.list.value[indexPath.item]
```
issueViewModel.list.value => . 이 두번 찍혔는데요! 한번만 찍고도 동일한 로직을 가져갈 수 있을까요?!!

Hansolkkim replied:
위에서 DataSource가 viewModel을 갖는게 맞냐는 질문에 답변드린 내용대로, issueViewModel말고 Observable<[IssueItem]> 타입의 속성을 해당 DataSource에서 사용하도록 변경될 예정이라,
`let issueEntity = data.value[indexPath.item]` 으로 `.`을 한 번만 찍도록 수정할 수 있을 것 같습니다!

아니면 아래와 같이 상수 하나를 추가로 두어, `.`의 갯수를 나누어보는 방향으로 수정도 가능할 것 같습니다.
```swift
let issueEntityList = issueViewModel.list
let issueEntity = issueEntityList.value[indexPath.item]
```

ITzombietux reviewed (approved):
Sol & Jason 이번 한주도 굉장히 수고많으셨습니다!! 벌써 프로젝트의 2/3가 흘렀네요!
6개월이라는 시간의 막바지에 날씨도 참 그렇고 여러모로 집중력이 흐려질 수 있는데도 학습과 구현을 잘 하시느라 너무나 수고하셨습니다. MVVM을 사용하며 VC의 역할을 잘 분리하려는 흔적이 너무나도 잘 드러나서 저도 많이 배우게 되는 시간이었습니다!! 주말 잘 쉬셔서 남은 한주 유종의 미를 거두시기를 기도할게요!

Hansolkkim committed:
refactor: NetworkTargetProtocol에 body 프로퍼티 추가 (bugpigg/issue-tracker#39)

- NetworkService의 makerequest 메소드에서 2중 if문을 제거하기 위해 프로퍼티 추가

Hansolkkim committed:
chore: 오타 수정 및 UI관련 작업을 main Thread에서 하도록 수정 (bugpigg/issue-tracker#39)

Hansolkkim committed:
refactor: DataSource 객체에서 가지고 있는 data의 타입 변경 (bugpigg/issue-tracker#39)

- dataSource에서는 단지 뷰의 속성을 변경해주기 위한 타입이 필요하기 때문에 viewModel을 인스턴스화 하여 속성으로 가지고 있을 필요가 없다 판단되어, viewModel이 아닌 Observable<[IssueItem]> 타입의 data를 가지고 있도록 수정

JasonLee0223 committed:
refactor: Default로 설정이 된 DataSource의 numberOfSections 메서드 코드 삭제

JasonLee0223 committed:
refactor: IssueItem의 Label 속성 중 darkMode 네이밍 변경으로 enum 추가

Hansolkkim committed:
refactor: -Managable 객체들의 setter 메소드 제거 (bugpigg/issue-tracker#39)

- urlSession의 setter 메소드는 urlSession을 public하게 놓는 것과 다를 것 없기 때문에, setter 메소드를 제거

JasonLee0223 committed:
refactor: update 메서드를 updateIssueView 메서드로 네이밍 변경

Hansolkkim committed:
Merge remote-tracking branch 'origin/iOS-refactor/ImplementFourthReview' into iOS

# Conflicts:
#	iOS/IssueTracker/IssueTracker/Issue/IssueCollectionViewDataSource.swift

JasonLee0223 committed:
refactor: issueItem darkMode CodingKey 다시 맞춤
